fix(compiler-cli): account for multiple generated namespace imports in HMR

crisbeto opened the pull request:
The current HMR compiler assumes that there will only be one namespace import in the generated code (`@angular/core`). This is incorrect, because the compiler may need to generate additional imports in some cases (e.g. importing directives through a module). These changes adjust the compiler to capture all the namespaces in an array and pass them along.

Fixes #58915.

JoostK commented on packages/compiler-cli/src/ngtsc/hmr/src/extract_dependencies.ts:
```diff
@@ -44,7 +43,15 @@ export function extractHmrLocals(
   // defined and it saves us having to repeat this logic wherever the locals are consumed.
   const availableTopLevel = getTopLevelDeclarationNames(sourceFile);
-  return Array.from(visitor.allReads).filter((r) => r !== name && availableTopLevel.has(r));
+  let idCounter = 0;
+
+  return {
+    local: Array.from(visitor.allReads).filter((r) => r !== name && availableTopLevel.has(r)),
+    external: Array.from(visitor.namespaceReads, (n) => ({
+      moduleName: n,
+      assignedName: `ɵhmr${idCounter++}`,
+    })),
+  };
```
I suppose this could use the `index` parameter from `Array,from` instead of maintaining a manual counter:

```suggestion
  return {
    local: Array.from(visitor.allReads).filter((r) => r !== name && availableTopLevel.has(r)),
    external: Array.from(visitor.namespaceReads, (n, index) => ({
      moduleName: n,
      assignedName: `ɵhmr${index}`,
    })),
  };
```

JoostK commented on packages/compiler-cli/src/ngtsc/hmr/src/update_declaration.ts:
```diff
@@ -52,12 +59,11 @@ export function getHmrUpdateDeclaration(
-/** Rewriter that replaces namespace imports to `@angular/core` with a specifier identifier. */
 class HmrModuleImportRewriter {
-  constructor(private readonly coreName: string) {}
+  constructor(private readonly lookup: Record<string, string>) {}
```
I'd change this to a `Map` as there's been frequent issues with incorrectly reading prototype properties.

JoostK commented on packages/compiler/src/render3/r3_hmr_compiler.ts:
```diff
@@ -43,15 +59,18 @@ export function compileHmrInitializer(meta: R3HmrMetadata): o.Expression {
   const defaultRead = o.variable(moduleName).prop('default');
 
-  // ɵɵreplaceMetadata(Comp, m.default, [...]);
+  // ɵɵreplaceMetadata(Comp, m.default, [...namespaces], [...locales]);
```
```suggestion
  // ɵɵreplaceMetadata(Comp, m.default, [...namespaces], [...locals]);
```

crisbeto commented:
👍  feedback has been addressed.

crisbeto committed:
fix(compiler-cli): account for multiple generated namespace imports in HMR

The current HMR compiler assumes that there will only be one namespace import in the generated code (`@angular/core`). This is incorrect, because the compiler may need to generate additional imports in some cases (e.g. importing directives through a module). These changes adjust the compiler to capture all the namespaces in an array and pass them along.

Fixes #58915.

pkozlowski-opensource commented:
This PR was merged into the repository by commit f280467398c6980878b5e755a78606251814447b.

The changes were merged into the following branches: main, 19.0.x
